lib/smack/SmackRep.cpp

```diff
@@ -758,36 +935,99 @@ const Expr* SmackRep::declareIsExternal(const Expr* e) {
-    s << "  havoc " << memReg(dstReg) << ";" << endl;
-    s << "  assume (forall x:int :: dest <= x && x < dest + len ==> "
-      << memReg(dstReg) << "[x] == $oldSrc[src - dest + x]);" << endl;
-    s << "  assume (forall x:int :: !(dest <= x && x < dest + len) ==> "
-      << memReg(dstReg) << "[x] == $oldDst[x]);" << endl;
-    s << "}" << endl;
+    if (SmackOptions::BitVectors) {
+      // TODO
+      s << "procedure $memcpy." << dstReg << "." << srcReg;
+      s << "(dest: ref, src: ref, len: ref, align: ref, isvolatile: bool)" << endl;
+      s << "modifies " << memPath(dstReg, 8) << ";" << endl;
+      if (SmackOptions::InferFieldOverlap) {
+	for (int i = 1; i < 4; ++i) {
+	  unsigned size = 8 << i;
+	  s << "modifies " << memPath(dstReg, size) << ";" << endl;
+	}
+      }
+      s << "{" << endl;
+      s << "  var $oldSrc" << ".i8" << ": [" << getPtrType() << "] " << int_type(8) << ";" << endl;
+      s << "  var $oldDst" << ".i8" << ": [" << getPtrType() << "] " << int_type(8) << ";" << endl;
+      if (SmackOptions::InferFieldOverlap) {
+	for (int i = 1; i < 4; ++i) {
+	  unsigned size = 8 << i;
+	  s << "  var $oldSrc" << ".i" << size << " : [" << getPtrType() << "] " << int_type(size) << ";" << endl;
+	  s << "  var $oldDst" << ".i" << size << " : [" << getPtrType() << "] " << int_type(size) << ";" << endl;
+	}
+      }
+      s << "  $oldSrc" << ".i8" << " := " << memPath(dstReg, 8) << ";" << endl;
+      s << "  $oldDst" << ".i8" << " := " << memPath(dstReg, 8) << ";" << endl;
+      s << "  havoc " << memPath(dstReg, 8) << ";" << endl;
+      if (SmackOptions::InferFieldOverlap) {
+	for (int i = 1; i < 4; ++i) {
+	  unsigned size = 8 << i;
+	  s << "  $oldSrc" << ".i" << size << " := " << memPath(srcReg, size) << ";" << endl;
+	  s << "  $oldDst" << ".i" << size << " := " << memPath(dstReg, size) << ";" << endl;
+	  s << "  havoc " << memPath(dstReg, size) << ";" << endl;
+	}
+      }
+      s << "  assume (forall x:i64 :: $ule.i32(dest, x) && $ult.i32(x, $add.i32(dest, len)) ==> "
+	<< memPath(dstReg, 8) << "[x] == $oldSrc.i8[$add.i32($sub.i32(src, dest), x)]);" << endl;
+      s << "  assume (forall x:i64 :: !($ule.i32(dest, x) && $ult.i32(x, $add.i32(dest, len))) ==> "
+	<< memPath(dstReg, 8) << "[x] == $oldDst.i8[x]);" << endl;
+      if (SmackOptions::InferFieldOverlap) {
+	for (int i = 1; i < 4; ++i) {
+	  unsigned size = 8 << i;
+	  s << "  assume (forall x:i64 :: $ule.i32(dest, x) && $ult.i32(x, $add.i32(dest, len)) ==> "
+	    << memPath(dstReg, size) << "[x] == $oldSrc" << ".i" << size << "[$add.i32($sub.i32(src, dest), x)]);" << endl;
+	  s << "  assume (forall x:i64 :: !($ule.i32(dest, x) && $ult.i32(x, $add.i32(dest, len))) ==> "
+	    << memPath(dstReg, size) << "[x] == $oldDst" << ".i" << size << "[x]);" << endl;
+	}
+      }
+      s << "}" << endl;
+    } else {
+      s << "procedure $memcpy." << dstReg << "." << srcReg;
+      s << "(dest: int, src: int, len: int, align: int, isvolatile: bool)" << endl;
```

Comment:
`dest` and `src` should have type `ref`, and `len` and `align` should have type `size`.